### PR TITLE
Reworks parameters again, fixes issue #288

### PR DESCRIPTION
**Pull request by jamonholmgren**
This removes a lot of the janky manual parsing of command line arguments that we were doing before and uses `minimist` for all of it.

My confidence level that this works across all situations is not super high, since I did rewrite a fair amount of critical path code here. Tests all pass, and in fact some jankiness with the tests is now looking much more like I'd expect, so that's a good sign.

### CODE REVIEW

File: src/domain/runtime.js [hunk 1]
```diff
@@ -60,32 +60,30 @@ async function run (rawCommand, options) {
     commandArray = commandArray.slice(2)
   }
 
-  // find the pluginName and commandName from the cli args
```
Comment (jamonholmgren) [inline src/domain/runtime.js]: Removed this whole block of code, using minimist via the `parseParams` function now.

File: src/domain/runtime.js [hunk 2]
```diff
@@ -94,15 +92,6 @@ async function run (rawCommand, options) {
     (this.defaults && this.defaults[context.plugin.name]) || {}
   )
 
-  // normalized parameters
```
Comment (jamonholmgren) [inline src/domain/runtime.js]: Not necessary, using the new `createParams` function above.

File: src/domain/runtime-parameters.test.js [hunk 3]
```diff
@@ -4,7 +4,7 @@ const Runtime = require('./runtime')
 test('can pass arguments', async t => {
   const r = new Runtime()
   r.load(`${__dirname}/../fixtures/good-plugins/args`)
-  const { command, parameters } = await r.run('args hello steve kellock', { caps: false })
```
Comment (jamonholmgren) [inline src/domain/runtime-parameters.test.js]: I never really liked how we had to put the plugin name here. It's no longer necessary.

I'm not _completely_ sure this is the right direction, though. I'm having some trouble wrapping my mind around the plugins command invocation and what @skellock intended here.

It does feel right, though.

File: src/domain/runtime-parameters.test.js [hunk 4]
```diff
@@ -32,3 +32,17 @@ test('can pass arguments, even with nested alias', async t => {
   t.deepEqual(parameters.options, { chocolate: true })
   t.deepEqual(command.commandPath, ['thing', 'foo'])
 })
+
+test('can pass arguments with mixed options', async t => {
+  const r = new Runtime()
+  r.load(`${__dirname}/../fixtures/good-plugins/args`)
+  const { command, parameters } = await r.run('--chocolate=true --foo -n 1 hello steve kellock')
```
Comment (jamonholmgren) [inline src/domain/runtime-parameters.test.js]: New regression test for #288 .

File: src/domain/runtime-parameters.test.js (hunk 3 above)
Comment (skellock) [reply src/domain/runtime-parameters.test.js]: The plugin doesn't make much sense in the new world order.  It's commands all the way down.

File: src/utils/cli/normalize-params.js [hunk 5]
```diff
@@ -1,20 +1,28 @@
 const minimist = require('minimist')
+const { merge } = require('ramda')
+
+function parseParams (argv, extraOpts = {}) {
+  // chop it up minimist!
+  const parsed = minimist(argv)
+  const array = parsed._.slice()
```
Comment (skellock) [inline src/utils/cli/normalize-params.js]: `._.` <--- awww... its so cute

**Comment by skellock** [comment]
One small thing i've noticed.

`am --force generate`

But these do:

```sh
am generate
am generate models
am generate --force
am generate models --force
am generate --force models
```

I don't think it's a deal breaker.  It'd be nice to support this though.  I wouldn't hold up this PR because of it though.

File: src/domain/runtime-parameters.test.js (hunk 3 above)
Comment (jamonholmgren) [reply src/domain/runtime-parameters.test.js]: I’m so glad you said that. It makes my life a lot easier. Ha

File: src/utils/cli/normalize-params.js (hunk 5 above)
Comment (jamonholmgren) [reply src/utils/cli/normalize-params.js]: Haha didn’t even notice that.

Comment (skellock) [reply src/utils/cli/normalize-params.js]: can't unsee

**Comment by jamonholmgren** [comment]
@skellock So this is a minimist issue, and essentially from what I can tell it's a design decision.

image

See the problem here? It's using the next arg as input. So it'll set `g` to the next argument. minimist wouldn't work with npm either. 😞 🐼 '

Given [this issue](https://github.com/substack/minimist/issues/115), and given that we _already had_ `yargs-parser` as a dependency (through a series of unfortunate events starting with `nyc`), I switched over to `yargs-parser` and removed `minimist`. yargs-parser exhibits the same exact behavior as above, so I think we will lean on the CLI creator to preprocess situations like `npm -g install` before handing it off to gluegun's runtime. Sound reasonable?

**Comment by jamonholmgren** [comment]
Semaphore is having issues, but tests are passing just fine.

**Comment by skellock** [comment]
`yargs-parser` wouldn't have been a production dependency tho because `nyc` is a dev dependency.

But this is all good with me.  `minimist` was internal anyway, so good call.

**Comment by jamonholmgren** [comment]
Oh, you're right. Didn't think about devDependencies. 👍

**Pull request merged**